File: sdk/compiler/daml-lf-proto-decode/src/DA/Daml/LF/Proto3/Decode.hs
-- Copyright (c) 2025 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
-- SPDX-License-Identifier: Apache-2.0

{-# LANGUAGE MultiWayIf #-}

module DA.Daml.LF.Proto3.Decode
  ( Error(..)
  , decodePayload
  , DecodeV2.decodePackage
  , DecodeV2.decodeSinglePackageModule
  ) where

import Com.Digitalasset.Daml.Lf.Archive.DamlLf (ArchivePayload(..), ArchivePayloadSum(..))

import qualified Proto3.Suite as Proto

import           DA.Daml.LF.Ast.Base     ( Package, PackageId, SelfOrImportedPackageId )
import           DA.Daml.LF.Proto3.Error
import qualified DA.Daml.LF.Proto3.DecodeV2 as DecodeV2
import qualified DA.Daml.LF.Ast             as LF
import           DA.Daml.StablePackagesList (stablePackages)

import           Control.Monad.Except (throwError)
import           Control.Lens         (over, _Left)
import Data.Int (Int32)
import qualified Data.Text.Lazy       as TL
import qualified Data.Text            as T

decodeLfVersion :: LF.MajorVersion -> LF.PackageId -> T.Text -> Int32 -> Either Error LF.Version
decodeLfVersion major pkgId minorText patchInt = do
  let unsupportedMinor :: Either Error a
      unsupportedMinor = throwError (UnsupportedMinorVersion minorText)
      unsupportedPatch :: LF.MinorVersion -> Int32 -> Either Error a
      unsupportedPatch minor patch = throwError (UnsupportedPatchVersion minor patch)
  minor <- if
    | Just minor <- LF.parseMinorVersion (T.unpack minorText) -> pure minor
    | otherwise -> unsupportedMinor
  _ <- if patchInt == 0 then pure () else unsupportedPatch minor patchInt
  let version = LF.Version major minor
  if pkgId `elem` stablePackages || version `elem` LF.supportedInputVersions
      then pure version
      else unsupportedMinor

decodePayload ::
    PackageId -> SelfOrImportedPackageId -> ArchivePayload -> Either Error Package
decodePayload pkgId selfPackageRef payload = case archivePayloadSum payload of
    Just (ArchivePayloadSumDamlLf1 _) -> do
         Left $  ParseError "Lf1 is not supported"
    Just (ArchivePayloadSumDamlLf2 packageBytes) -> do
        package <- over _Left (ParseError . show) $ Proto.fromByteString packageBytes
        version <- decodeLfVersion LF.V2 pkgId minorText patchInt
        DecodeV2.decodePackage version selfPackageRef package
    Nothing -> Left $ ParseError "Empty payload"
  where
    minorText = TL.toStrict (archivePayloadMinor payload)
    patchInt = archivePayloadPatch payload

File: sdk/daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Decode.scala
// Copyright (c) 2025 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.digitalasset.daml.lf.archive

import com.digitalasset.daml.lf.archive.DamlLf
import com.digitalasset.daml.lf.data.Ref.PackageId
import com.digitalasset.daml.lf.language.{Ast, LanguageMajorVersion, Util => AstUtil}

object Decode {

  def decodeArchivePayload(
      payload: ArchivePayload
  ): Either[Error, (PackageId, Ast.Package)] =
    decodeArchivePayload(payload, schemaMode = false)

  def decodeArchivePayloadSchema(
      payload: ArchivePayload
  ): Either[Error, (PackageId, Ast.PackageSignature)] =
    decodeArchivePayload(payload, schemaMode = true).map { case (pkgId, pkg) =>
      pkgId -> AstUtil.toSignature(pkg)
    }

  // decode an ArchivePayload
  private[this] def decodeArchivePayload(
      payload: ArchivePayload,
      schemaMode: Boolean,
  ): Either[Error, (PackageId, Ast.Package)] =
    payload match {
      case ArchivePayload.Lf2(pkgId, protoPkg, minor, patch)
          if LanguageMajorVersion.V2.supportedMinorVersions.contains(minor) =>
        new DecodeV2(minor)
          .decodePackage(
            pkgId,
            protoPkg,
            schemaMode,
            patch,
          )
          .map(payload.pkgId -> _)
      case ArchivePayload.Lf1(pkgId, protoPkg, minor)
          if LanguageMajorVersion.V1.supportedMinorVersions.contains(minor) =>
        new DecodeV1(minor)
          .decodePackage(
            pkgId,
            protoPkg,
            schemaMode,
          )
          .map(payload.pkgId -> _)
      case _ =>
        Left(Error.Parsing(s"${payload.version} unsupported"))
    }

  @throws[Error]
  def assertDecodeArchivePayload(
      payload: ArchivePayload,
      schemaMode: Boolean = false,
  ): (PackageId, Ast.Package) =
    assertRight(decodeArchivePayload(payload, schemaMode = schemaMode))

  // decode an Archive
  def decodeArchive(
      archive: DamlLf.Archive,
      schemaMode: Boolean = false,
  ): Either[Error, (PackageId, Ast.Package)] =
    Reader
      .readArchive(archive)
      .flatMap(decodeArchivePayload(_, schemaMode = schemaMode))

  @throws[Error]
  def assertDecodeArchive(
      archive: DamlLf.Archive,
      schemaMode: Boolean = false,
  ): (PackageId, Ast.Package) =
    assertRight(decodeArchive(archive, schemaMode = schemaMode))

}
